Documentation added for Spider State in extensions.rst (#6522)
scrapy/scrapy: docs/topics/extensions.rst

To enable this extension, turn on the :setting:`MEMDEBUG_ENABLED` setting. The
info will be stored in the stats.

Spider state extension
~~~~~~~~~~~~~~~~~~~~~~

.. module:: scrapy.extensions.spiderstate
   :synopsis: Spider state extension

.. class:: SpiderState

Manages spider state data by loading it before a crawl and saving it after.

Give a value to the :setting:`JOBDIR` setting to enable this extension.
When enabled, this extension manages the :attr:`~scrapy.Spider.state` 
attribute of your :class:`~scrapy.Spider` instance:
    
-   When your spider closes (:signal:`spider_closed`), the contents of its 
    :attr:`~scrapy.Spider.state` attribute are serialized into a file named 
    ``spider.state`` in the :setting:`JOBDIR` folder.
-   When your spider opens (:signal:`spider_opened`), if a previously-generated 
    ``spider.state`` file exists in the :setting:`JOBDIR` folder, it is loaded 
    into the :attr:`~scrapy.Spider.state` attribute.

> For an example, see :ref:`topics-keeping-persistent-state-between-batches`.

Reviewer: The text linked here says "a built-in extension", can you please make that a link to this text (by linking to `SpiderState`)?

To enable this extension, turn on the :setting:`MEMDEBUG_ENABLED` setting. The
info will be stored in the stats.

Spider state extension
~~~~~~~~~~~~~~~~~~~~~~

.. module:: scrapy.extensions.spiderstate

>    :synopsis: Close spider extension

Reviewer 2: This seems wrong.

.. class:: SpiderState

> Stores and loads spider state during a scraping job. It is especially useful when resuming a paused or interrupted job, as it ensures
> that the spider can retain its internal state, allowing for a more seamless continuation of the scraping process.
> 
> .. note:: To enable `SpiderState`, ensure that the `JOBDIR` setting is configured in your Scrapy project. Without this setting, the extension will raise a `NotConfigured` exception and will not be activated.
> 
> The extension is used while:
> 
> - **Opening a Spider**: When a spider is opened, the extension checks if a previous state file exists in the specified `JOBDIR`.
>   If found, it loads the state from the file and assigns it to `spider.state`.
> - **Closing a Spider**: When a spider closes, the extension saves the `spider.state` attribute to a file within `JOBDIR`.
>   This stored state can be reloaded in subsequent sessions, enabling continuity.

Reviewer 2: I think there are a few things to improve here, e.g.

- The way to enable the extension should not be a note, it should be a core part of the docs. And the fact that the extension is disabled with `NotConfigured` is an implementation detail, I don’t think it is useful here.
- There should be internal links to other relevant parts of the docs, such as the `Spider.state` reference docs. And we should probably link here from those docs as well.
- This is reStructuredText, not Markdown. A single ` makes text italic, not monospace. Please, use `tox -e docs` to build the docs locally, and open the result at `.tox/docs/tmp/html/index.html` to check that they render as you expect.

What about something like this?

```suggestion
Give a value to the :setting:`JOBDIR` setting to enable this extension.

When enabled, this extension manages the :attr:`~scrapy.Spider.state` 
attribute of your :class:`~scrapy.Spider` instance:
    
-   When your spider closes (:signal:`spider_closed`), the contents of its 
    :attr:`~scrapy.Spider.state` attribute are serialized into a file named 
    ``spider.state`` in the :setting:`JOBDIR` folder.

-   When your spider opens (:signal:`spider_opened`), if a previously-generated 
    ``spider.state`` file exists in the :setting:`JOBDIR` folder, it is loaded 
    into the :attr:`~scrapy.Spider.state` attribute.
```

> Implementation of state management in a spider:
> 
> .. code-block:: python
> 
>     from scrapy import Spider
> 
> 
>     class MySpider(Spider):
>         name = "my_spider"
>         # Additional setup...
> 
>         def parse(self, response):
>             # Use or update spider.state as needed
>             item = {}
>             self.state["last_item"] = item  # Storing state
>             return item

Reviewer 2: Rather than showing an unrealisting example here, I think it would be best to refer to the jobs docs for an example:

```suggestion
For an example, see :ref:`topics-keeping-persistent-state-between-batches`.
```

~~~~~~~~~~~~~~~~~~~~~~

.. module:: scrapy.extensions.spiderstate
   :synopsis: Spider state extension

.. class:: SpiderState

> Stores and loads spider state during a scraping job.

Reviewer 2: ```suggestion
Stores and loads spider state data during a scraping job.
```

“during a scraping job” feel misleading to me. It loads the spider state data *before* the crawl starts, and stores it *after* the crawl stops, it does not do anything during the crawl.

Author: How does this look for the intro "Manages spider state data by loading it before a crawl and saving it after."
